# Analysis of Used Car Sales in German Ebay Listings

In [1]:
import pandas as pd
import numpy as np

autos=pd.read_csv("autos.csv", encoding="Latin-1",)

print(autos.head())
autos.info()

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

Need to change data types of columns (price, odometer, dateCreated, lastSeen).  There are also various number of null values in the columns.

In [2]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Changing column names from camelCase to snake_case.

In [3]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [4]:
print(autos.describe(include='all'))

               date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  

Seller and offer_type columns have mostly the same values and the num_photos column needs to be investigated.  The odometer and price column values are type string instead of a numerical type (int or float).

In [5]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

All the values in the num_photos column are equal to zero and will need to be removed from analysis.

In [6]:
autos=autos.drop(["seller","offer_type","num_photos"], axis=1)
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

Removed seller, offer_type, and num_photos from dataframe.  Will now change datatypes of odometer and price columns to integers.  Also going to change datatype of postal_code to strings as postal codes cannot be added or substracted with itself in any meaningful way.  The columns with dates are in a string format and not date/time.

In [7]:
autos["odometer"]=autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
autos["odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [8]:
autos["price"]=autos["price"].str.replace(",","").str.replace("$","").astype(int)
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [9]:
autos["postal_code"]=autos["postal_code"].apply(str)
autos["postal_code"].head()

0    79588
1    71034
2    35394
3    33729
4    39218
Name: postal_code, dtype: object

In [10]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

# Exploring the Odometer and Price Columns

Now we will analyse the price and odometer columns for any obvious outliers.

In [11]:
autos["price"].unique().shape

(2357,)

In [12]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Min price is 0$ implying the vehicle(s) was given away.  Max price is 100,000,000$ which while possible needs to be verified due to how unlikely a vehicle of such worth would be sold on an eBay site.  For now best practice would be to remove the outliers. 

In [13]:
autos["price"].value_counts().head(10)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

1421 vehicles were given away.  The data is useless for analysis and will need to be removed.

In [14]:
autos["price"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

In [15]:
autos["price"][autos["price"]>1000000].count()

11

There are 11 vehicles that sold for over 1 million out of a total of 50,000 vehicles sold in the eBay Car Sales about 0.02%.

In [16]:
autos["price"][autos["price"]>1000000].describe()

count    1.100000e+01
mean     1.845509e+07
std      2.796876e+07
min      1.234566e+06
25%      6.945000e+06
50%      1.111111e+07
75%      1.234568e+07
max      1.000000e+08
Name: price, dtype: float64

Out of these 11 vehicles the price ranges from 1 million to 100 million.  These are outliers and can be excluded.

In [17]:
autos = autos[autos["price"].between(1,1000000)]
autos["price"].describe()

count     48568.000000
mean       5950.340656
std       11963.134750
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      999999.000000
Name: price, dtype: float64

Now we will analyze the odometer_km column.

In [18]:
autos["odometer_km"].value_counts()

150000    31416
125000     5058
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

Odometer values are all rounded implying the sellers had to pick from a preset drop down menu.  The values look reasonable and column will be left as is.

# Exploring Date Columns

As seen above, the data types of the columns containing specific dates/times are in a string format and not date/time.  First we will look at how the information is formatted. 

In [19]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We can extract the date by pulling the first 10 characters in each cell.

In [20]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48568, dtype: object


In [21]:
autos['date_crawled']=autos['date_crawled'].str[:10]
autos['date_crawled']=pd.to_datetime(autos["date_crawled"], format="%Y/%m/%d")
autos['date_crawled']

0       2016-03-26
1       2016-04-04
2       2016-03-26
3       2016-03-12
4       2016-04-01
           ...    
49995   2016-03-27
49996   2016-03-28
49997   2016-04-02
49998   2016-03-08
49999   2016-03-14
Name: date_crawled, Length: 48568, dtype: datetime64[ns]

In [22]:
autos['ad_created']=autos['ad_created'].str[:10]
autos['ad_created']=pd.to_datetime(autos["ad_created"], format="%Y/%m/%d")
autos['last_seen']=autos['last_seen'].str[:10]
autos['last_seen']=pd.to_datetime(autos["last_seen"], format="%Y/%m/%d")
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48568 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   date_crawled        48568 non-null  datetime64[ns]
 1   name                48568 non-null  object        
 2   price               48568 non-null  int64         
 3   ab_test             48568 non-null  object        
 4   vehicle_type        43982 non-null  object        
 5   registration_year   48568 non-null  int64         
 6   gearbox             46225 non-null  object        
 7   power_ps            48568 non-null  int64         
 8   model               46110 non-null  object        
 9   odometer_km         48568 non-null  int64         
 10  registration_month  48568 non-null  int64         
 11  fuel_type           44538 non-null  object        
 12  brand               48568 non-null  object        
 13  unrepaired_damage   39466 non-null  object    

The date columns are now in the correct format.  Specific times were removed to simplify analysis.

In [23]:
(autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025325
2016-03-06    0.014042
2016-03-07    0.036011
2016-03-08    0.033294
2016-03-09    0.033088
2016-03-10    0.032182
2016-03-11    0.032573
2016-03-12    0.036917
2016-03-13    0.015669
2016-03-14    0.036547
2016-03-15    0.034282
2016-03-16    0.029608
2016-03-17    0.031646
2016-03-18    0.012910
2016-03-19    0.034776
2016-03-20    0.037885
2016-03-21    0.037391
2016-03-22    0.032985
2016-03-23    0.032223
2016-03-24    0.029340
2016-03-25    0.031605
2016-03-26    0.032202
2016-03-27    0.031090
2016-03-28    0.034858
2016-03-29    0.034117
2016-03-30    0.033685
2016-03-31    0.031832
2016-04-01    0.033685
2016-04-02    0.035476
2016-04-03    0.038606
2016-04-04    0.036485
2016-04-05    0.013095
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [24]:
(autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_values())

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012910
2016-04-05    0.013095
2016-03-06    0.014042
2016-03-13    0.015669
2016-03-05    0.025325
2016-03-24    0.029340
2016-03-16    0.029608
2016-03-27    0.031090
2016-03-25    0.031605
2016-03-17    0.031646
2016-03-31    0.031832
2016-03-10    0.032182
2016-03-26    0.032202
2016-03-23    0.032223
2016-03-11    0.032573
2016-03-22    0.032985
2016-03-09    0.033088
2016-03-08    0.033294
2016-04-01    0.033685
2016-03-30    0.033685
2016-03-29    0.034117
2016-03-15    0.034282
2016-03-19    0.034776
2016-03-28    0.034858
2016-04-02    0.035476
2016-03-07    0.036011
2016-04-04    0.036485
2016-03-14    0.036547
2016-03-12    0.036917
2016-03-21    0.037391
2016-03-20    0.037885
2016-04-03    0.038606
Name: date_crawled, dtype: float64

The site was crawled for about a month and the distribution of submissions are roughly uniform.

In [25]:
(autos["ad_created"].value_counts(normalize=True, dropna=False).sort_values())

2015-12-30    0.000021
2016-01-03    0.000021
2015-08-10    0.000021
2016-01-16    0.000021
2016-01-07    0.000021
                ...   
2016-03-12    0.036753
2016-04-04    0.036856
2016-03-21    0.037597
2016-03-20    0.037947
2016-04-03    0.038853
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 2 months of the listing date, the oldest ads are about 8-10 months.

In [26]:
(autos["last_seen"].value_counts(normalize=True, dropna=False).sort_values())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005394
2016-03-18    0.007351
2016-03-08    0.007412
2016-03-13    0.008895
2016-03-09    0.009595
2016-03-10    0.010665
2016-03-11    0.012374
2016-03-14    0.012601
2016-03-27    0.015648
2016-03-19    0.015833
2016-03-15    0.015875
2016-03-16    0.016451
2016-03-26    0.016801
2016-03-23    0.018531
2016-03-25    0.019210
2016-03-24    0.019766
2016-03-21    0.020631
2016-03-20    0.020651
2016-03-28    0.020878
2016-03-22    0.021372
2016-03-29    0.022360
2016-04-01    0.022793
2016-03-12    0.023781
2016-03-31    0.023781
2016-04-04    0.024481
2016-03-30    0.024769
2016-04-02    0.024914
2016-04-03    0.025202
2016-03-17    0.028084
2016-04-05    0.124753
2016-04-07    0.131939
2016-04-06    0.221813
Name: last_seen, dtype: float64

The last_seen column refers to when the listing was last on the eBay site before being removed.  This could be because the vehicle was sold or removed by seller for whatever reason.  There is a presumed spike of sells from April 5 to April 7, 2016.

We will now look at the destribution of registration years.

In [27]:
autos["registration_year"].describe()

count    48568.000000
mean      2004.754612
std         88.641262
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There are some odd values in the registration_year column.  The earliest year is 1000ad before cars were invented and with a max year of 9999.

# Dealing with Incorrect Year Data

In [28]:
(autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.961209026519519

Over 96% of registration years are in reasonable timeframe.  We will remove the rows containing incorrect registration years.

In [29]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    46684.000000
mean      2002.910033
std          7.186122
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The distrubution now is more reasonable.  More than 50% of vehicles sold were registered after 2000.

# Exploring Price by Brand

Limiting analysis to just the top 20 brands.

In [30]:
top_20_brands=autos["brand"].value_counts(normalize=True).head(20).index
top_20_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [31]:
average_brand_price={}

for brand in top_20_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    
    average_brand_price[brand] = int(mean_price)

average_brand_price


{'volkswagen': 5604,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 4054,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397,
 'skoda': 6368,
 'nissan': 4743,
 'mazda': 4112,
 'smart': 3580,
 'citroen': 3779,
 'toyota': 5167,
 'hyundai': 5365,
 'sonstige_autos': 12338,
 'volvo': 4946,
 'mini': 10613}

In [32]:
bmp_series=pd.Series(average_brand_price).sort_values(ascending=False)
df=pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
sonstige_autos       12338
mini                 10613
audi                  9336
mercedes_benz         8628
bmw                   8332
skoda                 6368
volkswagen            5604
hyundai               5365
toyota                5167
volvo                 4946
nissan                4743
seat                  4397
mazda                 4112
ford                  4054
citroen               3779
smart                 3580
peugeot               3094
opel                  2975
fiat                  2813
renault               2474

The most expensive brands are above.  

# Top 20 Expensive Brands by Average Mileage

In [33]:
average_mean_mileage={}

for brand in top_20_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_km = brand_only["odometer_km"].mean()
    
    average_mean_mileage[brand] = int((mean_km)*0.621371)

amm_series=pd.Series(average_mean_mileage).sort_values(ascending=False)
df=pd.DataFrame(amm_series, columns=['odometer_miles'])
df

odometer_miles
volvo                    85791
bmw                      82376
mercedes_benz            81268
opel                     80349
audi                     80254
volkswagen               79977
renault                  79579
peugeot                  79009
mazda                    77338
ford                     77215
seat                     75267
citroen                  74374
nissan                   73527
fiat                     72776
toyota                   72044
skoda                    68878
hyundai                  66140
smart                    61718
sonstige_autos           55896
mini                     54745

Calculated top 20 brands by mileage.  Converted km to miles.

In [34]:
df["mean_price"] = bmp_series
df

odometer_miles  mean_price
volvo                    85791        4946
bmw                      82376        8332
mercedes_benz            81268        8628
opel                     80349        2975
audi                     80254        9336
volkswagen               79977        5604
renault                  79579        2474
peugeot                  79009        3094
mazda                    77338        4112
ford                     77215        4054
seat                     75267        4397
citroen                  74374        3779
nissan                   73527        4743
fiat                     72776        2813
toyota                   72044        5167
skoda                    68878        6368
hyundai                  66140        5365
smart                    61718        3580
sonstige_autos           55896       12338
mini                     54745       10613

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.